### PR TITLE
Preprocessnew [WiP]

### PR DESCRIPTION
**Pull request by david-novak**
New preprocessing widget.

### CODE REVIEW

File: orangecontrib/text/preprocess.py [hunk 1]
```diff
@@ -1,94 +1,161 @@
...
     """
-        Holds pre-processing flags and other information, about stop word
-        removal, lowercasing, text morphing etc.(the options are set via
-        the Preprocess widget).
+        A pre-processing class for the Orange text mining extension. An instance of the class
+        is capable of tokenizing, lowercasing, stemming and lemmatizing the input. Returns a list
+        of preprocessed tokens, sorted by order of appearance in text.
     """
-    def __init__(self, incl_punct=False, lowercase=True, stop_words='english', trans=None, min_df=1):
+    def __init__(self, tokenizer="word_tokenizer", lowercase=True, stop_words=None, transformation=None):
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: `word_tokenizer` seem quite long argument to write and remember. Couldn't we use just a flag insted? For example, `use_twitter_tokenize=False`?

File: orangecontrib/text/preprocess.py [hunk 2]
```diff
@@ -1,94 +1,161 @@
...
-        self.min_df = min_df
-        self.cv = CountVectorizer(
-            lowercase=lowercase,
-            stop_words=stop_words,
-            preprocessor=trans,
-            min_df=min_df,
-        )
+
+        # Stop words.
+        if stop_words == "default":
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: I suggest using `english` instead of `defualt`. Also, this cound be simplyfied to `self.stop_words = stop_words if stop_wrods != 'english' else stopwords.words('english')`

File: orangecontrib/text/preprocess.py [hunk 3]
```diff
@@ -1,94 +1,161 @@
...
+            self.stop_words = stop_words
+        elif stop_words is None:
+            self.stop_words = None
+        else:
+            raise ValueError("The stop words parameter should be either \'default\', "
+                             "a list containing the stop words or None.")
+
+        # Transformation.
+        self.transformation = transformation
+        if transformation == 'porter_stemmer':
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: Why don't we use the objects of type `Stemmatizer` directly? Then, all these following line cound be removed and we also remove the unnecessary mapping of these names to objects.

File: orangecontrib/text/preprocess.py [hunk 4]
```diff
@@ -1,94 +1,161 @@
...
+        """
+        Splits the text of the input into tokens by whitespaces and punctuation.
+        :param data: The input holding the text to be tokenized. Can be provided
+            in form of a string or list of strings.
+        :type data: string, list
+        :return: list
+        """
+        if isinstance(data, str):
+            return self.tokenizer(data)
+        elif isinstance(data, Orange.data.Table):
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: These three lines could safely be deleted until `Table`s are supported.

File: orangecontrib/text/preprocess.py [hunk 5]
```diff
@@ -1,94 +1,161 @@
...
+            # TODO: Handle this input.
+            return None
+        else:
+            raise ValueError("Cannot tokenize. Type {} not supported.".format(type(data)))
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: I suggest using just `Type {} not supported.`

File: orangecontrib/text/preprocess.py [hunk 6]
```diff
@@ -1,94 +1,161 @@
...
+            return self.transformation(str)
         elif isinstance(data, list):
-            if self.lowercase:
-                return [self.trans(word).lower() for word in data]
-            else:
-                return [self.trans(word) for word in data]
+            return [self.transformation(word) for word in data]
         else:
-            raise ValueError("Type {} not supported.".format(type(data)))
+            raise ValueError("Cannot transform. Type {} not supported.".format(type(data)))
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: I suggest using just `Type {} not supported.`

File: orangecontrib/text/preprocess.py [hunk 7]
```diff
@@ -1,94 +1,161 @@
...
+
+    def remove_stop_words(self, data):
+        """
+        Removes the stop words specified at class instance creation, from the list of tokens.
+        :param data: The list of string tokens, from where the stop words are to be removed.
+        :type data: list
+        :return: list
+        """
+        if self.stop_words is None:
+            raise ValueError("Cannot remove stop word, no set of such was specified.")
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: Prettify the error message.

File: orangecontrib/text/preprocess.py [hunk 8]
```diff
@@ -1,94 +1,161 @@
...
+    def transform(self, data):
+        """
+        Performs on the input, whatever transformation was specified at class instance creation.
+        :param data: The input holding the text to be transformed. Can be provided
+            in form of a string or list of string tokens.
+        :type data: string, list
+        :return: string, list
+        """
+        if self.transformation is None:
+            raise ValueError("Cannot perform transformation, none was specified.")
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: Prettify the error message.

File: orangecontrib/text/preprocess.py [hunk 9]
```diff
@@ -1,94 +1,161 @@
...
+
+    def transform(self, data):
+        """
+        Performs on the input, whatever transformation was specified at class instance creation.
+        :param data: The input holding the text to be transformed. Can be provided
+            in form of a string or list of string tokens.
+        :type data: string, list
+        :return: string, list
+        """
+        if self.transformation is None:
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: This is checked twice, here and in the `preprocess_document` method. Since this method is only called from the `preprocess_document` and is not meant to be called directly, we should check this only once. Also, fix the same duplication in `remove_stop_words` method.

File: orangecontrib/text/preprocess.py [hunk 10]
```diff
@@ -1,94 +1,161 @@
...
+
+import Orange
 
 
 class Preprocessor:
     """
-        Holds pre-processing flags and other information, about stop word
-        removal, lowercasing, text morphing etc.(the options are set via
-        the Preprocess widget).
+        A pre-processing class for the Orange text mining extension. An instance of the class
```
Comment (nikicc) [inline orangecontrib/text/preprocess.py]: `class for the Orange text mining extension. An instance of the class` can be deleted.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 11]
```diff
@@ -1,70 +1,1365 @@
...
+        """
+        start_file = os.path.expanduser("~/")
+        # Get the file path from the browser window.
+        path = QtGui.QFileDialog.getOpenFileName(
+            self, 'Open a stop words source', start_file, self.dlgFormats)
+        if not path:
+            return
+
+        new_files = [f for f in self.recent_sw_files if f != path]
+        #if len(new_files) < len(self.recent_sw_files):
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: Delete.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 12]
```diff
@@ -1,70 +1,1365 @@
...
+        new_files.insert(0, path)
+        self.recent_sw_files = new_files
+        self.set_file_list()
+        self.update_custom_source()
+
+    def open_file(self, path):
+        try:
+            self.current_file_path = path
+
+            #sw_list = self.read_sw_from_file(path)
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: Delete.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 13]
```diff
@@ -1,70 +1,1365 @@
...
+    outputs = [(Output.PP_CORPUS, Corpus)]
 
-    include_punctuation = Setting(False)
-    lowercase = Setting(True)
-    remove_stpwrds = Setting(True)
-    transformation_opt = Setting(["(none)", "Stemmer", "Lemmatizer"])
+    # Settings from previous runs. Actually chosen preprocessors.
+    stored_settings = settings.Setting({})
+    autocommit = settings.Setting(False)
+    recent_sw_files = settings.Setting(["/Users/david/Desktop/stopwords.txt"])
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: This should be empty! Clear custom settings.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 14]
```diff
@@ -1,70 +1,1365 @@
...
+
+        # Build a preprocessor instance with the selected parameters.
+        preprocessor = self.build_preproc()
+        if not preprocessor:
+            self.error(1, "No tokenizer specified.")
+            return
+
+        # Take the input corpus and append tokens to the metadata.
+        input_data = self.data.documents
+        output_corp = self.data
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: `output_corp ` is never used, delete it.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 15]
```diff
@@ -1,70 +1,1365 @@
...
+            params = pp_item.data(ParametersRole)
+
+            if not isinstance(params, dict):
+                params = {}
+
+            # Get the method from the BaseEditor class.
+            pp_info = desc.view_class.create_instance(params)
+
+            # Custom sw file. Read into list.
+            if "stop_words" in pp_info and pp_info["stop_words"] not in ["default", "None"]:
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: Reading stopword files shoud be done elsewhere. Now you read it even if some unrelated settings were modified. This shoud be done imediatly after the user selects the file from the dropdown on file system.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 16]
```diff
@@ -1,70 +1,1365 @@
...
+        self.data.tokens = tokens
+        self.send(Output.PP_CORPUS, self.data)
+
+        # Sync the model into stored_settings on every apply.
+        self.storeSpecificSettings()
+
+    def build_preproc(self):
+        self.warning(1, "")
+        # Preprocessor parameters that we require.
+        p = {"tokenizer": None, "lowercase": False, "stop_words": None, "transformation": None}
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: The scripting part alredy has the defults set, why do we need to set them once again here?

File: orangecontrib/text/widgets/owpreprocess.py [hunk 17]
```diff
@@ -1,70 +1,1365 @@
...
+                    self.warning(1, "Stop word remover: {}".format(err))
+
+            p.update(pp_info)
+        print("Pre-procesing with: \n", p)
+
+        # Return false if no tokenizer was specified.
+        if p["tokenizer"] is None:
+            return False
+
+        return Preprocessor(tokenizer=p["tokenizer"], lowercase=p["lowercase"], stop_words=p["stop_words"],
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: Simplify this to `Preprocessor(**pp_info)`.

File: orangecontrib/text/widgets/owpreprocess.py [hunk 18]
```diff
@@ -1,70 +1,1365 @@
...
+
+            # Get the method from the BaseEditor class.
+            pp_info = desc.view_class.create_instance(params)
+
+            # Custom sw file. Read into list.
+            if "stop_words" in pp_info and pp_info["stop_words"] not in ["default", "None"]:
+                sw_files = pp_info["stop_words"]
+                try:
+                    with open(pp_info["stop_words"][0]) as f:
+                        pp_info["stop_words"] = f.readlines()
```
Comment (nikicc) [inline orangecontrib/text/widgets/owpreprocess.py]: Please strip the `\n` characters from stop words if words are presented in multiple lines.

**Comment by nikicc** [comment]
I propose the following modifications and/or bug fixes:
* All `prints` and commented lines of code (some were found in `owpreprocessor.py`) should be deleted.
* When `None` is selected in `Stop word remover` it causes a unnecessary `NoneType object is not subscriptable` warning.
* `Lowercase` checkbox IMO fits better into the `Stemmer` section than into the `Tokenizer`.
* Until `twitter tokenizer` is implemented, please gray/comment it out from the GUI.
* I suggest removing the `tokenizer type` and `stemmer type` labels from the GUI, since the title of the section tells what this is about.
* Should we rename the `word tokenizer` option to `NLTK tokenizer` or something similar, since `twitter tokenizer` is also a word tokenizer, just adopted for tweets?
* The sections could be renamed to `Tokenization`, `Stemming` or even better `Normalization`, and `Filtering` or `Token Filtering`.
* It would be awesome if we implemented progress bar, since Lemmatization can take quite some time.
* Selecting different stopwords file does not trigger a commit when autocommit is enabled.
* A reload button for stopword files would be very nice.
* We could enhance the `Stopword removal` with the option to remove tokens by document frequency`min_df` and `max_df` which are either `float` of `int` like in [scikit](http://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html).

About the order of preprocessors it might be better if tokenizer is always on top. Also, since we support different order of `Stemmer` and `Stopword removal` this should either be implemented in scripting part (to support custom order) or we should disable the support for custom ordering in the widget to prevent confusion.

**Comment by nikicc** [comment]
Tests fail due to the following line in `test_preprocess.py`: `from orangecontrib.text.preprocess import Stemmer` since `Stemmer` was replaced with `PorterStemmer` and `SnowballStemmer`. Could you also incorporate this along with the other fixes?

File: orangecontrib/text/preprocess.py [hunk 19]
```diff
@@ -1,94 +1,161 @@
...
+        is capable of tokenizing, lowercasing, stemming and lemmatizing the input. Returns a list
+        of preprocessed tokens, sorted by order of appearance in text.
     """
-    def __init__(self, incl_punct=False, lowercase=True, stop_words='english', trans=None, min_df=1):
+    def __init__(self, tokenizer="word_tokenizer", lowercase=True, stop_words=None, transformation=None):
         """
-        :param incl_punct: Determines whether the tokenizer should include punctuation in the tokens.
-        :type incl_punct: boolean
+        :param tokenization: A mandatory argument that defines the type of tokenizer we wish to use.
```
Comment (kernc) [inline orangecontrib/text/preprocess.py]: While having a go at all the @nikicc's suggestions, would you also consider documenting methods in [Napoleon (NumPy) style](http://sphinxcontrib-napoleon.readthedocs.org/en/latest/example_numpy.html#example-numpy)? They render beautifully in ipython and are easier to look at in clear text. :smile:

File: orangecontrib/text/preprocess.py (hunk 2 above)
Comment (kernc) [inline orangecontrib/text/preprocess.py]: NLTK has some 2400 stopwords for 11 languages. I'd leave the `list` of words case and the `'language'` case handled separately.

File: orangecontrib/text/preprocess.py (hunk 9 above)
Comment (kernc) [inline orangecontrib/text/preprocess.py]: Better safe than sorry. `transform()`, not starting with an underscore, conveys a public interface.

File: orangecontrib/text/preprocess.py (hunk 2 above)
Comment (david-novak) [inline orangecontrib/text/preprocess.py]: On the matter of separating these parameters, how about we introduce the stop words list as an 'additional_stop_words' parameter? Then the user can choose the NLTK stop words and specify a list of custom sw. Or just use the custom list, setting the NLTK option to None.

Comment (kernc) [inline orangecontrib/text/preprocess.py]: If the user can extend the built-in nltk list simply by concatenation, `stopwords.words('english') + stopwords.words('japanese')`, I'm against it. Another parameter makes for less simple and quite redundant interface.

Comment (nikicc) [inline orangecontrib/text/preprocess.py]: I also wrote for just one parameter.

File: orangecontrib/text/widgets/owpreprocess.py (hunk 16 above)
Comment (david-novak) [inline orangecontrib/text/widgets/owpreprocess.py]: The dict is used only because we can update it gradually with every iteration. We could update the Preprocessor class, so we can update individual parameters, but for now it can only be created through the constructor, which demands all parameters at once.

**Comment by nikicc** [comment]
The tests are failing due to missing NLTK data. Please, add the required packages to `travis/nltk_download.sh` in line `6`.

**Comment by nikicc** [comment]
@david-novak we can close this now, right?

**Comment by david-novak** [comment]
@nikicc Yes, everything seems to be in order.
